pkg/maintainer/spv/spv.go

```diff
@@ -57,10 +98,302 @@ func (sm *spvMaintainer) startControlLoop(ctx context.Context) {
+	requiredConfirmations, err := sm.chain.TxProofDifficultyFactor()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get transaction proof difficulty factor: [%v]",
+			err,
+		)
+	}
+
+	for _, transaction := range depositSweepTransactions {
+		// TODO: Skip if the transaction has not accumulated enough confirmations yet.
+
+		_, proof, err := bitcoin.AssembleSpvProof(
+			transaction.Hash(),
+			uint(requiredConfirmations.Uint64()),
+			sm.btcChain,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to assemble SPV proof: [%v]", err)
+		}
+
+		mainUTXO, vault, err := parseTransactionInputs(
+			sm.btcChain,
+			sm.chain,
+			transaction,
+		)
+		if err != nil {
+			return fmt.Errorf(
+				"error while parsing transaction inputs: [%v]",
+				err,
+			)
+		}
+
+		// TODO: Remember that the relay may temporarily be in the out-of-date
+		//       state. It happens at the beginning of each Bitcoin difficulty
+		//       epoch. Detect the situation when the proof contains block
+		//       headers with a difficulty that is not yet proven. Skip proving
+		//       such a transaction. It will be proven in the future by another
+		//       round of processing deposit sweep proofs.
+		if err := sm.chain.SubmitDepositSweepProofWithReimbursement(
+			transaction,
+			proof,
+			mainUTXO,
+			vault,
+		); err != nil {
+			return fmt.Errorf(
+				"failed to submit deposit sweep proof with reimbursement: [%v]",
+				err,
+			)
+		}
+	}
 
 	return nil
 }
+
+func (sm *spvMaintainer) getUnprovenDepositSweepTransactions() (
+	[]*bitcoin.Transaction,
+	error,
+) {
+	blockCounter, err := sm.chain.BlockCounter()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get block counter: [%v]", err)
+	}
+
+	currentBlock, err := blockCounter.CurrentBlock()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current block: [%v]", err)
+	}
+
+	// Calculate the starting block of the range in which the events will be
+	// searched for.
+	startBlock := currentBlock - sm.config.HistoryDepth
+
+	depositSweepTransactionProposals, err :=
+		sm.chain.PastDepositSweepProposalSubmittedEvents(
+			&tbtc.DepositSweepProposalSubmittedEventFilter{
+				StartBlock: startBlock,
+			},
+		)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"failed to get past deposit sweep proposal submitted events: [%v]",
+			err,
+		)
+	}
+
+	// There will often be multiple events emitted for a single wallet. Prepare
+	// a list of unique wallet public key hashes.
+	walletPublicKeyHashes := uniqueWalletPublicKeyHashes(
+		depositSweepTransactionProposals,
+	)
+
+	unprovenDepositSweepTransactions := []*bitcoin.Transaction{}
+
+	for _, walletPublicKeyHash := range walletPublicKeyHashes {
+		wallet, err := sm.chain.GetWallet(walletPublicKeyHash)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get wallet: [%v]", err)
+		}
+
+		// Skip if the wallet is not `Live` or `MovingFunds`.
+		if wallet.State != 1 && wallet.State != 2 {
+			continue
+		}
```

Comment:
The values of `1` and `2` are Solidity/EVM implementation details. I think we should modify `WalletChainData` structure to return `State` not as `uint8` but as `WalletState` enum we will introduce. We have `WalletActionType` and `parseWalletActionType` in the Ethereum implementation. We should introduce identical enum for `WalletState` and have an identical parsing function. I don't think we should necessarily reuse the same code given `WalletActionType` comes from the `WalletCoordinator` contract and it is another type.

Reply:
Done in 2386f5ba3d386d0db1144f88dc4b76631dc9901b.
Unfortunately, I had to prefix each state constant:
```
const (
	StateUnknown WalletState = iota
	StateLive
	StateMovingFunds
	StateClosing
	StateClosed
	StateTerminated
)
``` 
with `State...`, because we already have `MovingFunds` constant in `WalletActionType`.

Reply:
I propose that we prefix all types of `WalletActionType` with `Action*` in a separate PR to keep it consistent.

Reply:
Done in a separate PR https://github.com/keep-network/keep-core/pull/3662.